File: Biblioteca-main/Movil/biblioteca_flutter/lib/services/notification_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'api_service.dart';
import 'auth_service.dart';
import '../models/notificacion.dart';
import 'package:intl/intl.dart';

class NotificationService {
  final AuthService _authService = AuthService();

  static const int POLLING_INTERVAL = 30; // segundos
  static const int MULTA_POR_DIA = 1000; // $1000 por día

  Timer? _pollingTimer;
  final _notificacionesController =
      StreamController<List<Notificacion>>.broadcast();

  Stream<List<Notificacion>> get notificaciones$ =>
      _notificacionesController.stream;

  Set<String> _notificacionesVistas = {};
  List<Notificacion> _notificacionesDevolucion = [];

  NotificationService() {
    _cargarDesdePreferencias();
  }

  // Cargar notificaciones vistas y de devolución desde SharedPreferences
  Future<void> _cargarDesdePreferencias() async {
    final prefs = await SharedPreferences.getInstance();

    // Cargar vistas
    final vistasJson = prefs.getString('notificaciones_vistas');
    if (vistasJson != null) {
      final List<dynamic> vistas = jsonDecode(vistasJson);
      _notificacionesVistas = vistas.map((e) => e.toString()).toSet();
    }

    // Cargar devoluciones
    final devolucionJson = prefs.getString('notificaciones_devolucion');
    if (devolucionJson != null) {
      final List<dynamic> devolucion = jsonDecode(devolucionJson);
      _notificacionesDevolucion = devolucion
          .map((json) => Notificacion.fromJson(json as Map<String, dynamic>))
          .toList();
    }
  }

  // Calcular días de retraso
  int _calcularDiasRetraso(DateTime fechaHasta) {
    final ahora = DateTime.now();
    final diferencia = ahora.difference(fechaHasta).inDays;
    return diferencia > 0 ? diferencia : 0;
  }

  // Formatear fecha
  String _formatearFecha(DateTime fecha) {
    final formato = DateFormat('dd/MM/yyyy HH:mm');
    return formato.format(fecha);
  }

  // Crear mensaje para vencida
  String _crearMensajeVencida(Notificacion notif) {
    final fechaDesdeStr = _formatearFecha(notif.fechaDesde!);
    final fechaHastaStr = _formatearFecha(notif.fechaHasta!);

    return 'Buenos días, estimado/a usuario/a, le queremos comentar que su reserva del libro "${notif.libroTitulo}" necesita ser devuelto, según los datos de fecha "$fechaDesdeStr - $fechaHastaStr", por estos inconvenientes, se le aplicará una multa de ${notif.diasRetraso} día(s) equivalente a \$${notif.multaEconomica}';
  }

  // Crear mensaje para devolución
  String _crearMensajeDevolucion(Notificacion notif) {
    final fechaStr = _formatearFecha(notif.fechaDevolucion!);
    return 'Libro "${notif.libroTitulo}" devuelto con éxito y pagado el $fechaStr con multa de \$${notif.multaEconomica}';
  }

  // Obtener notificaciones (vencidas + devoluciones)
  Future<List<Notificacion>> fetchNotificaciones() async {
    final token = await _authService.getToken();
    if (token == null) return [..._notificacionesDevolucion];

    try {
      final response = await ApiService.get('/reservas/vencidas', token: token);

      if (response['success'] == true) {
        final data = response['data'];

        // Notificaciones vencidas (filtrar las vistas)
        List<Notificacion> notifVencidas = [];
        if (data is List) {
          notifVencidas = data
              .where(
                  (reserva) => !_notificacionesVistas.contains(reserva['_id']))
              .map((reserva) {
            final diasRetraso =
                _calcularDiasRetraso(DateTime.parse(reserva['hasta']));
            final notif = Notificacion(
              id: reserva['_id'] ?? '',
              tipo: 'vencida',
              libroTitulo:
                  reserva['libro_id']?['titulo'] ?? 'Libro desconocido',
              fechaDesde: DateTime.parse(reserva['desde']),
              fechaHasta: DateTime.parse(reserva['hasta']),
              diasRetraso: diasRetraso,
              multa: diasRetraso,
              multaEconomica: diasRetraso * MULTA_POR_DIA,
              mensaje: '',
            );

            return Notificacion(
              id: notif.id,
              tipo: notif.tipo,
              libroTitulo: notif.libroTitulo,
              fechaDesde: notif.fechaDesde,
              fechaHasta: notif.fechaHasta,
              diasRetraso: notif.diasRetraso,
              multa: notif.multa,
              multaEconomica: notif.multaEconomica,
              mensaje: _crearMensajeVencida(notif),
            );
          }).toList();
        }

        // Combinar con notificaciones de devolución
        final todasNotificaciones = [
          ...notifVencidas,
          ..._notificacionesDevolucion
        ];
        _notificacionesController.add(todasNotificaciones);
        return todasNotificaciones;
      }
    } catch (e) {
      print('Error obteniendo notificaciones: $e');
    }

    // En caso de error, retornar solo las de devolución
    _notificacionesController.add([..._notificacionesDevolucion]);
    return [..._notificacionesDevolucion];
  }

  // Marcar como vista
  Future<void> marcarComoVista(String id) async {
    if (id.startsWith('dev_')) {
      // Notificación de devolución: eliminar
      _notificacionesDevolucion.removeWhere((n) => n.id == id);
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(
          'notificaciones_devolucion',
          jsonEncode(
              _notificacionesDevolucion.map((n) => n.toJson()).toList()));
    } else {
      // Notificación vencida: marcar como vista
      _notificacionesVistas.add(id);
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(
          'notificaciones_vistas', jsonEncode(_notificacionesVistas.toList()));
    }

    await fetchNotificaciones();
  }

  // Crear notificación de devolución
  Future<void> crearNotificacionDevolucion(
    String libroTitulo,
    int multa,
    int multaEconomica,
  ) async {
    final notif = Notificacion(
      id: 'dev_${DateTime.now().millisecondsSinceEpoch}',
      tipo: 'devolucion',
      libroTitulo: libroTitulo,
      fechaDevolucion: DateTime.now(),
      multa: multa,
      multaEconomica: multaEconomica,
      mensaje: '',
    );

    final notifConMensaje = Notificacion(
      id: notif.id,
      tipo: notif.tipo,
      libroTitulo: notif.libroTitulo,
      fechaDevolucion: notif.fechaDevolucion,
      multa: notif.multa,
      multaEconomica: notif.multaEconomica,
      mensaje: _crearMensajeDevolucion(notif),
    );

    _notificacionesDevolucion.add(notifConMensaje);

    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('notificaciones_devolucion',
        jsonEncode(_notificacionesDevolucion.map((n) => n.toJson()).toList()));

    await fetchNotificaciones();
  }

  // Iniciar polling
  void iniciarPolling() {
    if (_pollingTimer != null) return;

    // Primera carga inmediata
    fetchNotificaciones();

    // Polling periódico
    _pollingTimer = Timer.periodic(
      Duration(seconds: POLLING_INTERVAL),
      (_) => fetchNotificaciones(),
    );
  }

  // Detener polling
  void detenerPolling() {
    _pollingTimer?.cancel();
    _pollingTimer = null;
  }

  // Limpiar datos al cerrar sesión
  Future<void> limpiarDatos() async {
    _notificacionesVistas.clear();
    _notificacionesDevolucion.clear();

    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('notificaciones_vistas');
    await prefs.remove('notificaciones_devolucion');

    _notificacionesController.add([]);
  }

  void dispose() {
    detenerPolling();
    _notificacionesController.close();
  }
}

File: Biblioteca-main/src/app/services/notification.service.ts
import { Injectable, OnDestroy, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { BehaviorSubject, Observable, interval, Subscription } from 'rxjs';
import { switchMap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';

export interface Notificacion {
    id: string;
    tipo: 'vencida' | 'devolucion';
    libroTitulo: string;
    fechaDesde?: Date;
    fechaHasta?: Date;
    fechaDevolucion?: Date;
    diasRetraso?: number;
    multa: number;
    multaEconomica: number;
    mensaje: string;
}

@Injectable({
    providedIn: 'root'
})
export class NotificationService implements OnDestroy {
    private apiUrl = 'http://127.0.0.1:8000/api';
    private notificacionesSubject = new BehaviorSubject<Notificacion[]>([]);
    public notificaciones$ = this.notificacionesSubject.asObservable();

    private pollingSubscription?: Subscription;
    private readonly POLLING_INTERVAL = 30000; // 30 segundos
    private readonly MULTA_POR_DIA = 1000; // $1000 por día de retraso

    private notificacionesVistas: Set<string> = new Set();
    private notificacionesDevolucion: Notificacion[] = [];
    private isBrowser: boolean;

    constructor(
        private http: HttpClient,
        @Inject(PLATFORM_ID) platformId: object
    ) {
        this.isBrowser = isPlatformBrowser(platformId);

        if (this.isBrowser) {
            // Cargar notificaciones vistas desde localStorage
            const vistas = localStorage.getItem('notificaciones_vistas');
            if (vistas) {
                this.notificacionesVistas = new Set(JSON.parse(vistas));
            }

            // Cargar notificaciones de devolución
            const devolucion = localStorage.getItem('notificaciones_devolucion');
            if (devolucion) {
                this.notificacionesDevolucion = JSON.parse(devolucion).map((n: any) => ({
                    ...n,
                    fechaDevolucion: n.fechaDevolucion ? new Date(n.fechaDevolucion) : undefined
                }));
            }
        }
    }

    // Obtener headers con token
    private getHeaders(): { headers: HttpHeaders } {
        const token = localStorage.getItem('auth_token');
        return {
            headers: new HttpHeaders({
                'Authorization': `Bearer ${token}`,
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            })
        };
    }

    // Calcular días de retraso
    private calcularDiasRetraso(fechaHasta: Date): number {
        const ahora = new Date();
        const diffTime = ahora.getTime() - new Date(fechaHasta).getTime();
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        return diffDays > 0 ? diffDays : 0;
    }

    // Formatear fecha
    private formatearFecha(fecha: Date): string {
        const f = new Date(fecha);
        const dia = String(f.getDate()).padStart(2, '0');
        const mes = String(f.getMonth() + 1).padStart(2, '0');
        const anio = f.getFullYear();
        const horas = String(f.getHours()).padStart(2, '0');
        const minutos = String(f.getMinutes()).padStart(2, '0');
        return `${dia}/${mes}/${anio} ${horas}:${minutos}`;
    }

    // Crear mensaje formateado para vencidas
    private crearMensajeVencida(notif: Notificacion): string {
        const fechaDesdeStr = this.formatearFecha(notif.fechaDesde!);
        const fechaHastaStr = this.formatearFecha(notif.fechaHasta!);

        return `Buenos días, estimado/a usuario/a, le queremos comentar que su reserva del libro "${notif.libroTitulo}" necesita ser devuelto, según los datos de fecha "${fechaDesdeStr} - ${fechaHastaStr}", por estos inconvenientes, se le aplicará una multa de ${notif.diasRetraso} día(s) equivalente a $${notif.multaEconomica}`;
    }

    // Crear mensaje para devolución
    private crearMensajeDevolucion(notif: Notificacion): string {
        const fechaStr = this.formatearFecha(notif.fechaDevolucion!);
        return `Libro "${notif.libroTitulo}" devuelto con éxito y pagado el ${fechaStr} con multa de $${notif.multaEconomica}`;
    }

    // Obtener reservas vencidas y convertirlas en notificaciones
    private fetchNotificaciones(): Observable<Notificacion[]> {
        return this.http.get<any[]>(`${this.apiUrl}/reservas/vencidas`, this.getHeaders()).pipe(
            switchMap(reservas => {
                // Notificaciones de libros vencidos (filtrar las vistas)
                const notifVencidas: Notificacion[] = reservas
                    .filter(reserva => !this.notificacionesVistas.has(reserva._id))
                    .map(reserva => {
                        const diasRetraso = this.calcularDiasRetraso(reserva.hasta);
                        const notif: Notificacion = {
                            id: reserva._id,
                            tipo: 'vencida',
                            libroTitulo: reserva.libro_id?.titulo || 'Libro desconocido',
                            fechaDesde: reserva.desde,
                            fechaHasta: reserva.hasta,
                            diasRetraso: diasRetraso,
                            multa: diasRetraso,
                            multaEconomica: diasRetraso * this.MULTA_POR_DIA,
                            mensaje: ''
                        };
                        notif.mensaje = this.crearMensajeVencida(notif);
                        return notif;
                    });

                // Combinar con notificaciones de devolución
                const todasNotificaciones = [...notifVencidas, ...this.notificacionesDevolucion];

                this.notificacionesSubject.next(todasNotificaciones);
                return of(todasNotificaciones);
            }),
            catchError(error => {
                console.error('Error obteniendo notificaciones:', error);
                // Aunque falle el backend, mostrar notificaciones de devolución
                this.notificacionesSubject.next([...this.notificacionesDevolucion]);
                return of([...this.notificacionesDevolucion]);
            })
        );
    }

    // Marcar notificación como vista/cerrada
    marcarComoVista(id: string): void {
        if (!this.isBrowser) return;

        // Si es notificación de devolución, eliminarla completamente
        if (id.startsWith('dev_')) {
            this.notificacionesDevolucion = this.notificacionesDevolucion.filter(n => n.id !== id);
            localStorage.setItem('notificaciones_devolucion', JSON.stringify(this.notificacionesDevolucion));
        } else {
            // Si es notificación vencida, marcarla como vista
            this.notificacionesVistas.add(id);
            localStorage.setItem('notificaciones_vistas', JSON.stringify([...this.notificacionesVistas]));
        }

        // Actualizar la lista
        this.fetchNotificaciones().subscribe();
    }

    // Crear notificación de devolución
    crearNotificacionDevolucion(libroTitulo: string, multa: number, multaEconomica: number): void {
        if (!this.isBrowser) return;

        const notif: Notificacion = {
            id: `dev_${Date.now()}`,
            tipo: 'devolucion',
            libroTitulo: libroTitulo,
            fechaDevolucion: new Date(),
            multa: multa,
            multaEconomica: multaEconomica,
            mensaje: ''
        };

        notif.mensaje = this.crearMensajeDevolucion(notif);

        this.notificacionesDevolucion.push(notif);
        localStorage.setItem('notificaciones_devolucion', JSON.stringify(this.notificacionesDevolucion));

        // Actualizar inmediatamente
        this.fetchNotificaciones().subscribe();
    }

    // Iniciar polling automático
    iniciarPolling(): void {
        if (this.pollingSubscription) {
            return; // Ya está activo
        }

        // Primera carga inmediata
        this.fetchNotificaciones().subscribe();

        // Polling periódico
        this.pollingSubscription = interval(this.POLLING_INTERVAL).pipe(
            switchMap(() => this.fetchNotificaciones())
        ).subscribe();
    }

    // Detener polling
    detenerPolling(): void {
        if (this.pollingSubscription) {
            this.pollingSubscription.unsubscribe();
            this.pollingSubscription = undefined;
        }
        this.notificacionesSubject.next([]);
    }

    // Limpiar datos al cerrar sesión
    limpiarDatos(): void {
        if (!this.isBrowser) return;

        this.notificacionesVistas.clear();
        this.notificacionesDevolucion = [];
        localStorage.removeItem('notificaciones_vistas');
        localStorage.removeItem('notificaciones_devolucion');
    }

    // Obtener el número de notificaciones
    getContador(): Observable<number> {
        return new Observable(observer => {
            this.notificaciones$.subscribe(notificaciones => {
                observer.next(notificaciones.length);
            });
        });
    }

    ngOnDestroy(): void {
        this.detenerPolling();
    }
}
